Make GTK an optional dependency

rayment opened the pull request:
I thought I might as well make a PR for this given I've been running it since February without issue:

I run Gentoo without GTK and compiling it to run DF was a pain so I took direct inspiration from [rofl0r's repo](https://github.com/rofl0r/df-libgraphics), which mind you, hasn't been updated in years, and made GTK optional, replacing it with terminal input instead if it isn't present on build.

Included in the PR is also a bugfix from [here](https://www.bay12games.com/dwarves/mantisbt/view.php?id=11564) which fixes a few functions that weren't returning anything and causing crashes on Gentoo (and quite possibly other systems).

svenstaro commented:
Could you rebase on master? I just pushed a basic CI.

svenstaro commented on CMakeLists.txt:
```diff
@@ -14,10 +14,14 @@ find_package(Freetype REQUIRED)
+set(CURSES_NEED_NCURSES FALSE)
+find_package(Curses)
+
+if(GTK_FOUND)
+   target_compile_definitions(graphics PRIVATE HAVE_GTK)
+endif()
```
I think you'll have to move this down to the very bottom of the file as otherwise you'll get
```
CMake Error at CMakeLists.txt:23 (target_compile_definitions):
  Cannot specify compile definitions for target "graphics" which is not built
  by this project.
```

rayment replied:
Strange, I don't get this error (I'm on CMake 3.20.5). I do however get an error in `g_src/renderer_curses.cpp` claiming that `mvwaddstr` is not defined. I'll move it down anyway.

svenstaro commented on g_src/win32_compat.cpp:
```diff
@@ -141,6 +146,17 @@ int MessageBox(HWND *dummy, const char *text, const char *caption, UINT type)
         break;
       }
     }
+  #endif
+    dprintf(2, "Alert %s:\n%s\n", caption ? caption : "", text ? text : "");
+    if (type & MB_YESNO) {
+      while(ret == IDOK) {
+        dprintf(2, "please answer with 'yes' or 'no'\n");
+        char buf[16];
+        fgets(buf, sizeof buf, stdin);
+        if(!strncmp(buf, "yes", 3)) ret = IDYES;
+        else if(!strncmp(buf, "no", 2)) ret = IDNO;
+      }
+    }
```
Not quite sure I understand this. Why is this useful to have?

rayment replied:
This comes from rofl0r's repository, it provides terminal dialog when ncurses and GTK are not available. I have to check I have the defines right though.

svenstaro commented:
While I don't really have the need for this, I do see how some people would. Happy to accept this, thanks, but I think it's broken in its current state.

rayment commented:
You're right, I apologise for the sloppy PR. I'm getting build errors too now on my Gentoo, I'll resolve this and report back.

rayment commented:
Alright so I've moved the GTK check to the bottom of `CMakeLists.txt` and added an extra little if statement that ensures windowed mode is chosen automatically if no dialog can be spawned (no GTK present and no terminal is open).

Please verify that this is okay.

rayment commented:
Pushed one last commit to ensure the terminal won't send a request as well as opening the GTK dialog. I think that should do it now.

rayment reviewed (commented):
Reviewed changes, see last comment.

svenstaro commented:
Alright, seems good now. Thanks!
